docs/connection/connection-spans.md

```diff
@@ -0,0 +1,180 @@
+
+| Value  | Description |
+|---|---|
+| `tcp` | TCP |
+| `udp` | UDP |
+| `pipe` | Named or anonymous pipe. |
+| `unix` | Unix domain socket |
+
+`network.type` has the following list of well-known values. If one of them applies, then the respective value MUST be used, otherwise a custom value MAY be used.
+
+| Value  | Description |
+|---|---|
+| `ipv4` | IPv4 |
+| `ipv6` | IPv6 |
+<!-- endsemconv -->
+
+## Examples
+
+### Successful connection
+
+Successful connection attempt to `"/tmp/my.sock"` results in the following span:
+
+|   Attribute name       |         Value       |
+| :--------------------- | :-------------------|
+| name                   | `"connect"`         |
+| `network.peer.address` | `"/tmp/my.sock"`    |
+| `network.transport`    | `"unix"`            |
+
+Once corresponding connection is gracefully closed, another span is reported:
+
+|   Attribute name       |         Value       |
+| :--------------------- | :-------------------|
+| name                   | `"connection"`      |
+| `network.peer.address` | `"/tmp/my.sock"`    |
+| `network.transport`    | `"unix"`            |
+
+### Successful connect, but connection terminates with an error
+
+Successful connection attempt to `example.com` results in the following span:
+> Note: DNS lookup is outside of the scope of this semantic convention
+
+|   Attribute name       |         Value       |
+| :--------------------- | :-------------------|
+| name                   | `"connect"`         |
+| `server.address`       | `"example.com"`     |
+| `network.peer.address` | `"93.184.216.34"`   |
+| `network.peer.port`    | `443`               |
+| `network.transport`    | `"tcp"`             |
+| `network.transport`    | `"ipv4"`            |
+
+But then after some packet exchange, the connection is reset:
+
+|   Attribute name       |         Value       |
+| :--------------------- | :-------------------|
+| name                   | `"connection"`      |
+| `server.address`       | `"example.com"`     |
+| `network.peer.address` | `"93.184.216.34"`   |
+| `network.peer.port`    | `443`               |
+| `network.transport`    | `"tcp"`             |
+| `network.transport`    | `"ipv4"`            |
+| `error.type`           | `econnreset`        |
+
+### Attempt to establish connection ends with `econnrefused` error
+
+An attempt to establish connection to `127.0.0.1:8080` without any application
+listening on this port results in the following span:
+
+|   Attribute name       |         Value       |
+| :--------------------- | :-------------------|
+| name                   | `"connect"`         |
+| `network.peer.address` | `"127.0.0.1"`   |
+| `network.peer.port`    | `8080`               |
+| `network.transport`    | `"tcp"`             |
+| `network.type`         | `"ipv4"`            |
+| `error.type`           | `econnrefused`      |
+
+### Relationship with application protocols such as HTTP
```

Comment:
Http becomes interesting with connection pooling, and the ability to either do sequential (http 1.1) or parallel (http2,3) requests over the same connection.
We would probably want some form or event for when the wire-requests are put on to a specific connection. Similarly for DNS lookup and TLS handshake, those are events that occur as part of connection establishment that are important to collect in some way for deeper diagnostics. Should they be events as defined by tracing, or log messages tagged with the same traceid/spanid?

Reply:
why events? we have links to correlate request to connection and we can put attributes on them if any is necessary.

Do you want to capture moment in time when the request is associated with the connection? We don't capture it on links yet, but we can start. Record a link and an event is an overkill.

I wonder if DNS and TLS should be spans or events. Since they are network calls with non-zero duration spans would work better (but will be slightly less performant).
Given that connections live much longer than requests, the volume of such spans would be low and perf should not be a big deal.

Reply:
I am thinking that this needs to be a dial that ops can turn based on how much data that they want to collect. Using the scenario of an HttpClient call (outgoing):
- You have HttpClient spans (implemented today) - these are really tracking a logical request rather than what is physically happening on the wire.
- The next level would be a chain of physical requests - in the case of redirection, it could be multiple before the final request, or it could do a continue to resume collection of data.
- The connections themselves are longer lived beyond a single request. They have DNS and TLS as part of the initialization.
- DNS lookup for a connection
- TLS negotiation

In most cases you probably don't want to collect all of these all the time. However I can see ops turning them on when needed to collect more specific diagnostics data. So can we make it adaptive, and be able to correlate the data when applicable?

I am wondering if an "event" + optional link approach would be best. When a request is put on the wire for a connection - you'd get an event - that way you kind of know what the delay was before your request was processed. If the connections are being tracked, then that "event" would have a link to the connection span, so you could correlate them together.
Similarly a connection would have an "event" for when the dns resolution has occurred, and TLS negotiation is complete. If either are being tracked, that event would link to their respective spans - although those could probably be parented to the connection.

I use "event" in quotes as I am told the future of events on spans is unclear - it could be done with a log message instead.